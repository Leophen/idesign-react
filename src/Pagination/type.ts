export interface PageInfo {
  current: number;
  pageSize: number;
}

export interface PaginationProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 当前页
   * @default 1
   */
  current?: number;
  /**
   * 数据总条数
   * @default 0
   */
  total?: number
  /**
   * 分页总页数
   * @default 10
   */
  pageSize?: number;
  /**
   * 全局禁用
   * @default false
   */
  disabled?: boolean;
  /**
   * 是否隐藏分页选择器
   * @default false
   */
  hideSelect?: boolean;
  /**
   * 是否隐藏跳转输入框
   * @default false
   */
  hideInput?: boolean;
  /**
   * 当前页或分页发生变化时触发
   */
  onChange?: (pageInfo: PageInfo) => void;
}

export interface PaginationSelectProps {
  /**
   * 每页条数
   * @default 10
   */
  value?: number
  /**
   * 禁用状态
   * @default false
   */
  disabled?: boolean;
  /**
   * 选择每页条数时触发
   */
  onSelect?: (val: number) => void
}

export interface PaginationListProps extends PaginationProps {
  /**
   * 点击页数时触发
   */
  onChoose?: (val: number) => void
}
