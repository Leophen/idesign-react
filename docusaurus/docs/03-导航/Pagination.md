---
title: Pagination 分页
---

展现的信息行数过多时，可采用分页控制单页内的信息数量，也可帮助用户快速到定位某一内容的位置。

## 引入

```js
import {Pagination} from 'idesign';
```

## 基本用法

```jsx live fffx
<Pagination defaultCurrent={1} total={50} />
```

## 页数较多

```jsx live fffx
<Pagination total={505} defaultCurrent={5} />
```

## 快速跳转

快速跳转到某一页。

```jsx live fffx
function App() {
    function onChange(pageNumber) {
        console.log('Page: ', pageNumber);
    }

    function showTotal(total) {
        return `共 ${total} 条`;
    }

    return (
        <div className="demo-column">
            <Pagination
                showQuickJumper
                defaultCurrent={2}
                total={500}
                onChange={onChange}
                showTotal={showTotal}
            />
            <Pagination
                showQuickJumper
                defaultCurrent={2}
                total={500}
                onChange={onChange}
                disabled
                showTotal={showTotal}
            />
        </div>
    );
}
```

## 每页条目数

改变每页显示条目数。

```jsx live fff
function App() {
    const [current, setCurrent] = useState(1);
    const [pageSize, setPageSize] = useState(20);

    function onShowSizeChange(current, pageSize) {
        console.log(current, pageSize);
    }

    function showTotal(total) {
        return `共 ${total} 条`;
    }

    return (
        <div className="demo-column">
            <Pagination
                showSizeChanger
                onShowSizeChange={onShowSizeChange}
                defaultCurrent={1}
                total={500}
                showTotal={showTotal}
                onChange={(page, pageSize) => {
                    console.log(page, pageSize);
                }}
            />
            <Pagination
                showSizeChanger
                onShowSizeChange={onShowSizeChange}
                defaultCurrent={1}
                total={500}
                showTotal={showTotal}
                disabled
            />
            <Pagination
                showSizeChanger
                current={current}
                pageSize={pageSize}
                total={505}
                pageSizeOptions={[10, 20, 50, 100]}
                onChange={(page, pageSize) => {
                    console.log(page, pageSize);
                    setCurrent(page);
                    setPageSize(pageSize);
                }}
            />
        </div>
    );
}
```

## API

| 属性             | 说明                                                     | 类型                               | 默认值              |
| ---------------- | -------------------------------------------------------- | ---------------------------------- | ------------------- |
| current          | 当前页数                                                 | `number`                           | `--`                |
| defaultCurrent   | 默认的当前页数                                           | `number`                           | `1`                 |
| defaultPageSize  | 默认的每页条数                                           | `number`                           | `10`                |
| disabled         | 禁用分页                                                 | `boolean`                          | `--`                |
| hideOnSinglePage | 只有一页时是否隐藏分页器                                 | `boolean`                          | `false`             |
| itemRender       | 用于自定义页码的结构，可用于优化 SEO                     | `(page, type: 'page'`              | `'prev'`            |
| pageSize         | 每页条数                                                 | `number`                           | `--`                |
| pageSizeOptions  | 指定每页可以显示多少条                                   | `string[]`                         | `[10, 20, 50, 100]` |
| responsive       | 当 size 未指定时，根据屏幕宽度自动调整尺寸               | `boolean`                          | `--`                |
| showLessItems    | 是否显示较少页面内容                                     | `boolean`                          | `false`             |
| showQuickJumper  | 是否可以快速跳转至某页                                   | `boolean〡{ goButton: ReactNode }` | `false`             |
| showSizeChanger  | 是否展示 pageSize 切换器，当 total 大于 50 时默认为 true | `boolean`                          | `--`                |
| showTitle        | 是否显示原生 tooltip 页码提示                            | `boolean`                          | `true`              |
| showTotal        | 用于显示数据总量和当前数据顺序                           | `function(total, range)`           | `--`                |
| simple           | 当添加该属性时，显示为简单分页                           | `boolean`                          | `--`                |
| size             | 当为 small 时，是小尺寸分页                              | `default〡small`                   | `default`           |
| total            | 数据总数                                                 | `number`                           | `0`                 |
| onChange         | 页码改变的回调，参数是改变后的页码及每页条数             | `function(page, pageSize)`         | `--`                |
| onShowSizeChange | pageSize 变化的回调                                      | `function(current, size)`          | `--`                |
