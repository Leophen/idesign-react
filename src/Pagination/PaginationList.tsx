import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import Icon from '../Icon';
import _ from 'lodash';
import { PaginationListProps } from './type';

const PaginationList: React.FC<PaginationListProps> = (props) => {
  const {
    current = 1,
    total = 0,
    pageSize = 10,
    disabled = false,
    onChoose
  } = props

  const [pageArr, setPageArr] = useState(_.range(1, _.ceil(total / pageSize) + 1))

  const handleChooseNum = (item: number) => {
    if (!disabled) {
      let result = item
      result < 1 && (result = 1)
      result > pageArr.length && (result = pageArr.length)
      onChoose?.(result)
    }
  }

  const shortItems = (
    pageArr.map(item => (
      <div
        className={classNames(
          'i-pagination-btn',
          current === item && 'i-pagination-btn__active',
          disabled && 'i-pagination-btn__disabled'
        )}
        key={`item${item}`}
        onClick={() => handleChooseNum(item)}
      >
        {item}
      </div>
    ))
  )

  const updateArrowShow = () => {
    let lShow = true
    let rShow = true
    if (_.indexOf(pageArr, current) < 4) {
      lShow = false
    } else if (_.indexOf(pageArr, current) > pageArr.length - 5) {
      rShow = false
    }
    setLeftShow(lShow)
    setRightShow(rShow)
  }

  const pagesRef = useRef<HTMLElement>(null)

  useEffect(() => {
    updateArrowShow()
    pagesRef.current && pagesRef.current.scrollTo({
      left: 32 * (current - 3) + 8 * (current - 3),
      behavior: 'smooth'
    });
    const newCountArr = _.range(1, _.ceil(total / pageSize) + 1)
    setPageArr(newCountArr)
  }, [current, total, pageSize])

  const [leftShow, setLeftShow] = useState(false)
  const [rightShow, setRightShow] = useState(false)

  const [leftHover, setLeftHover] = useState(false)
  const [rightHover, setRightHover] = useState(false)

  const longItems = (
    <>
      {leftShow && (
        <div
          className={classNames(
            'i-pagination-btn',
            '-handle',
            disabled && 'i-pagination-btn__disabled'
          )}
          onMouseEnter={() => setLeftHover(true)}
          onMouseLeave={() => setLeftHover(false)}
          onClick={disabled ? () => { } : () => handleChooseNum(current - 5)}
        >
          <Icon name={leftHover ? "ArrowDoubleLeft" : "MoreFilled"} size={12} />
        </div>
      )}
      <section className="i-pagination-btn__wrapper" ref={pagesRef}>
        {pageArr.map(item => (
          <div
            className={classNames(
              'i-pagination-btn',
              current === item && 'i-pagination-btn__active',
              disabled && 'i-pagination-btn__disabled'
            )}
            key={`item${item}`}
            onClick={() => handleChooseNum(item)}
          >
            {item}
          </div>
        ))}
      </section>
      {rightShow && (
        <div
          className={classNames(
            'i-pagination-btn',
            '-handle',
            disabled && 'i-pagination-btn__disabled'
          )}
          onMouseEnter={() => setRightHover(true)}
          onMouseLeave={() => setRightHover(false)}
          onClick={disabled ? () => { } : () => handleChooseNum(current + 5)}
        >
          <Icon name={rightHover ? "ArrowDoubleRight" : "MoreFilled"} size={12} />
        </div>
      )}
    </>
  )

  return (
    <div
      className="i-pagination-control"
    >
      <div
        className={classNames(
          'i-pagination-btn',
          'i-pagination-btn__handle',
          (current === 1 || disabled) && 'i-pagination-btn__disabled'
        )}
        onClick={() => current !== 1 && !disabled && onChoose?.(current - 1)}
      >
        <Icon name="ArrowLeft" size={12} />
      </div>
      {pageArr.length <= 10 ? shortItems : longItems}
      <div
        className={classNames(
          'i-pagination-btn',
          'i-pagination-btn__handle',
          (current === pageArr.length || disabled) && 'i-pagination-btn__disabled'
        )}
        onClick={() => current !== pageArr.length && !disabled && onChoose?.(current + 1)}
      >
        <Icon name="ArrowRight" size={12} />
      </div>
    </div>
  )
}

export default PaginationList;
