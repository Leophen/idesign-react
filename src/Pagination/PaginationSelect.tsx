import React from 'react';
import './index.scss';
import Select from '../Select';
import { PaginationSelectProps } from './type';

const PaginationSelect: React.FC<PaginationSelectProps> = (props) => {
  const {
    value = 10,
    disabled = false,
    onSelect
  } = props

  const options = [
    {
      content: '10 条/页',
      value: 10
    },
    {
      content: '20 条/页',
      value: 20
    },
    {
      content: '30 条/页',
      value: 30
    },
    {
      content: '40 条/页',
      value: 40
    },
    {
      content: '50 条/页',
      value: 50
    }
  ]

  const handleChange = (val: string | number | Array<string | number>) => {
    onSelect?.(val as number)
  }

  return (
    <div className="i-pagination-select">
      <Select
        value={value}
        disabled={disabled}
        options={options}
        clearable={false}
        onChange={handleChange}
      />
    </div>
  )
}

export default PaginationSelect;
