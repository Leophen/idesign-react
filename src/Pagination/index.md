---
nav:
  title: 组件
  path: /components
group:
  title: 导航组件
  order: 2
order: 4
---

# Pagination 分页

用于在当前页切换内容的分页组件。

## 基本用法

```tsx
import React from 'react';
import { Pagination } from 'idesign-react';

const App = () => {
  const handleChange = (val) => {
    console.log(val)
  }

  return (
    <div className="idesign-demo-block-column">
      <Pagination total={100} />
      <Pagination current={10} total={999} pageSize={20} onChange={handleChange} />
    </div>
  );
};

export default App;
```

## 全局禁用

可通过 `disabled` 属性隐藏分页选择器：

```tsx
import React from 'react';
import { Pagination } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <Pagination disabled total={100} />
      <Pagination disabled current={10} total={999} pageSize={20} />
    </div>
  );
};

export default App;
```

## 隐藏分页选择器

可通过 `hideSelect` 属性隐藏分页选择器：

```tsx
import React from 'react';
import { Pagination } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <Pagination hideSelect total={100} />
      <Pagination hideSelect current={10} total={999} pageSize={20} />
    </div>
  );
};

export default App;
```

## 隐藏跳转输入框

可通过 `hideInput` 属性隐藏分页选择器：

```tsx
import React from 'react';
import { Pagination } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <Pagination hideInput total={100} />
      <Pagination hideSelect hideInput current={10} total={999} pageSize={20} />
    </div>
  );
};

export default App;
```

<API />
