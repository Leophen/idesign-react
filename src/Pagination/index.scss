.i-pagination {
  display: flex;
  gap: 16px;
  align-items: center;
}

.i-pagination-select {
  width: 100px;
}

.i-pagination-control {
  display: flex;
  gap: 8px;
  align-items: center;
}

.i-pagination-btn {
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  width: 32px;
  min-width: 32px;
  height: 32px;
  color: var(--i-font-2);
  font-size: 14px;
  border: 1px solid var(--i-border);
  border-radius: 4px;
  cursor: pointer;
  user-select: none;
  &:hover {
    color: var(--i-primary-txt);
    border-color: var(--i-primary-txt);
  }
  &.i-pagination-btn__handle {
    border: none;
    &:hover {
      background: var(--i-bg-hover);
    }
  }
  &.i-pagination-btn__active {
    color: var(--i-bg);
    background: var(--i-primary-txt);
    border-color: var(--i-primary-txt);
    &:hover {
      color: var(--i-bg);
    }
  }
  &.i-pagination-btn__disabled {
    color: var(--i-font-disabled);
    cursor: not-allowed;
    &:hover {
      color: var(--i-font-disabled);
      background: unset;
      border-color: var(--i-border);
    }
    &.i-pagination-btn__handle {
      border: none;
    }
    &.i-pagination-btn__active {
      color: var(--i-bg);
      background: var(--i-primary-disabled);
      border-color: var(--i-primary-disabled);
    }
  }
}

.i-pagination-btn__wrapper {
  display: flex;
  gap: 8px;
  align-items: center;
  width: 192px;
  overflow: hidden;
  &::-webkit-scrollbar {
    display: none;
  }
}

.i-pagination-input {
  display: flex;
  gap: 8px;
  align-items: center;
  color: var(--i-font-2);
  font-size: 14px;
}
