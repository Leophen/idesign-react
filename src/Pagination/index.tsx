import React, { useEffect, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import Input from '../Input';
import _ from 'lodash';
import PaginationSelect from './PaginationSelect';
import PaginationList from './PaginationList';
import { PaginationProps } from './type';

const Pagination: React.FC<PaginationProps> = (props) => {
  const {
    className,
    style,
    current = 1,
    total = 0,
    pageSize = 10,
    disabled = false,
    hideSelect = false,
    hideInput = false,
    onChange,
    ...restProps
  } = props;

  const [pageData, setPageData] = useState({
    current: current,
    total: total,
    pageSize: pageSize,
  })

  const switchPageSize = (pageSize: number) => {
    console.log(current, total, pageSize)
    pageData.pageSize = pageSize
    setPageData({ ...pageData })
    onChange?.(pageData)
  }

  const clickPageNum = (page: number) => {
    pageData.current = page
    setPageData({ ...pageData })
    onChange?.(pageData)
  }

  const [inputVal, setInputVal] = useState(String(pageData.current))

  useEffect(() => {
    setInputVal(String(pageData.current))
  }, [pageData.current])

  const handleInputEnter = (val: string) => {
    if (val !== '') {
      clickPageNum(Number(val))
    } else {
      clickPageNum(1)
    }
  }

  const handleInputChange = (val: string) => {
    setInputVal(val)
  }

  return (
    <div
      className={classNames(
        'i-pagination',
        className
      )}
      style={{ ...style }}
      {...restProps}
    >
      {!hideSelect && (
        <PaginationSelect
          value={pageData.pageSize}
          disabled={disabled}
          onSelect={switchPageSize}
        />
      )}
      <PaginationList
        current={pageData.current}
        total={pageData.total}
        pageSize={pageData.pageSize}
        disabled={disabled}
        onChoose={clickPageNum}
      />
      {!hideInput && (
        <section className='i-pagination-input'>
          <span>前往</span>
          <Input
            style={{
              width: String(_.ceil(pageData.total / pageData.pageSize)).length * 8 + 20
            }}
            placeholder=''
            type='number'
            speed='slow'
            hideNumberBtn
            value={inputVal}
            disabled={disabled}
            maxNumber={_.ceil(pageData.total / pageData.pageSize)}
            minNumber={1}
            onChange={handleInputChange}
            onEnter={handleInputEnter}
            onMoveUp={handleInputEnter}
          />
          <span>页</span>
        </section>
      )}
    </div>
  );
};

Pagination.displayName = 'Pagination';

export default Pagination;
