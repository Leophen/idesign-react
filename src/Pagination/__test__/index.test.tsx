import { render } from '@testing-library/react';
import Pagination from '../index';

describe('Pagination 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<Pagination total={100} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

