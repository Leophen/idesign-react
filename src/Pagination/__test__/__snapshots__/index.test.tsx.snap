// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-pagination"
  >
    <div
      class="i-pagination-select"
    >
      <div
        class="i-select"
      >
        <div
          class="i-popup__reference"
        >
          <div
            class="i-input i-input-is-readonly i-input__hide-clear"
          >
            <input
              class="i-input__inner"
              placeholder="请选择"
              readonly=""
              value="10 条/页"
            />
            <i
              class="i-icon icon-TipCloseFill"
            />
            <i
              class="i-icon icon-ArrowDown i-input-suffix-icon"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="i-pagination-control"
    >
      <div
        class="i-pagination-btn i-pagination-btn__handle i-pagination-btn__disabled"
      >
        <i
          class="i-icon icon-ArrowLeft"
          style="font-size: 12px;"
        />
      </div>
      <div
        class="i-pagination-btn i-pagination-btn__active"
      >
        1
      </div>
      <div
        class="i-pagination-btn"
      >
        2
      </div>
      <div
        class="i-pagination-btn"
      >
        3
      </div>
      <div
        class="i-pagination-btn"
      >
        4
      </div>
      <div
        class="i-pagination-btn"
      >
        5
      </div>
      <div
        class="i-pagination-btn"
      >
        6
      </div>
      <div
        class="i-pagination-btn"
      >
        7
      </div>
      <div
        class="i-pagination-btn"
      >
        8
      </div>
      <div
        class="i-pagination-btn"
      >
        9
      </div>
      <div
        class="i-pagination-btn"
      >
        10
      </div>
      <div
        class="i-pagination-btn i-pagination-btn__handle"
      >
        <i
          class="i-icon icon-ArrowRight"
          style="font-size: 12px;"
        />
      </div>
    </div>
    <section
      class="i-pagination-input"
    >
      <span>
        前往
      </span>
      <div
        class="i-input i-input--type-number"
        style="width: 36px;"
      >
        <input
          class="i-input__inner"
          max="10"
          min="1"
          placeholder=""
          step="1"
          type="number"
          value="1"
        />
        <div
          class="i-input-number-slider"
        />
      </div>
      <span>
        页
      </span>
    </section>
  </div>
</DocumentFragment>
`;
